aten/src/ATen/native/cpu/IndexKernel.cpp

```diff
@@ -131,6 +131,38 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
   });
 }
 
+void index_fill_kernel(
+  TensorIterator& iter,
+  int64_t dim,
+  int64_t self_dim_size,
+  int64_t self_dim_stride,
+  Scalar source) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16,
+    iter.dtype(), "index_fill_cpu", [&] {
+    auto fill_val = source.to<scalar_t>();
+    auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+      auto* self_data_bytes = data[0];
+      auto* index_data_bytes = data[1];
+      for (int64_t elem = 0; elem < n; ++elem) {
+        auto* self_data = reinterpret_cast<scalar_t*>(self_data_bytes);
+        auto idx = *reinterpret_cast<int64_t*>(index_data_bytes);
+
+        if (idx < -self_dim_size || idx >= self_dim_size) {
+          TORCH_CHECK_INDEX(false,
+            "index ", idx, " is out of bounds for dimension ",
+            dim, " with size ", self_dim_size);
+        }
+
+        self_data[idx * self_dim_stride] = fill_val;
+
+        self_data_bytes += strides[0];
+        index_data_bytes += strides[1];
```

Comment:
I got an idea that maybe if strides[1] is 0 (which is the case for indexing over non-last dimension) you should not be doing this, and instead get `idx` value outside of `elem` loop, and then just either move self_data_bytes forward or use self_data with increasing index. That might help compiler auto-vectorize this loop -currently it can't be done because `idx` is read every time, and compiler has no way of knowing that it will be the same.

Reply:
Yep, taking `idx` out of the loop significantly improves perf for 1000,1000 case on my system, 3.7s->2.7s

Reply:
Also valid index check for this case can be taken outside the loop, for additional perf gains.

Reply:
Done, indeed, now it is faster!